R/download_link.R

```diff
@@ -0,0 +1,130 @@
+#' @param file_size The file size if known. Needs to be a string ending in one
+#' of KB, MB, GB or rows.
+#'
+#' @returns shiny.tag object
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' ui <- fluidPage(
+#'   p("Choose a dataset to download."),
+#'   selectInput("dataset", "Dataset", choices = c("mtcars", "airquality")),
+#'   download_link("download_data", "Download data")
+#' )
+#'
+#' server <- function(input, output) {
+#'   # The requested dataset
+#'   data <- reactive({
+#'     get(input$dataset)
+#'   })
+#'
+#'   output$download_data <- downloadHandler(
+#'     filename = function() {
+#'       # Use the selected dataset as the suggested file name
+#'       paste0(input$dataset, ".csv")
+#'     },
+#'     content = function(file) {
+#'       # Write the dataset to the `file` that will be downloaded
+#'       write.csv(data(), file)
+#'     }
+#'   )
+#' }
+#'
+#' shinyApp(ui, server)
+#' }
+download_link <- function(
+    outputId,
+    link_text = "Download data",
+    file_type = "CSV",
+    file_size = NULL) {
+  # Trim white space as I don't trust humans not to accidentally include
+  link_text <- stringr::str_trim(link_text)
+
+  # Create a basic check for raw URLs
+  is_url <- function(text) {
+    url_pattern <- "^(https://|http://|www\\.)"
+    grepl(url_pattern, text)
+  }
+
+  # Check for vague link text on our list
+  if (is_url(link_text)) {
+    stop(paste0(
+      link_text,
+      " has been recognised as a raw URL, please change the link_text value",
+      "to a description of the page being linked to instead"
+    ))
+  }
+
+  # Check against curated data set for link text we should banish into room 101
+  if (tolower(link_text) %in% shinyGovstyle::bad_link_text$bad_link_text) {
+    stop(
+      paste0(
+        link_text,
+        " is not descriptive enough and has has been recognised as bad link",
+        " text, please replace the link_text argument with more descriptive",
+        " text."
+      )
+    )
+  }
+
+  # Check if link text ends in a full stop
+  if (grepl("\\.$", link_text)) {
+    stop("link_text should not end with a full stop")
+  }
+
+    # Give a console warning if link text is under 7 characters
+  # Arbritary number that allows for R Shiny to be link text without a warning
+  if (nchar(link_text) < 7) {
+    warning(paste0(
+      "the link_text: ", link_text, ", is shorter than 7 characters, this is",
+      " unlikely to be descriptive for users, consider having more detailed",
+      " link text"
+    ))
+  }
+
+  # Assuming all else has passed, make the link text a nice accessible link
+  # Note the file_size is manually assigned and optional right now. Ideally,
+  # this would be updated dynamically when linking to a dynamically created
+  # file, such as the CSV version of a table in an app.
+  if (!is.null(file_size)) {
+    if(!grepl("KB$|MB$|GB$| rows$", file_size)){
+      error("File size should be a string ending in one of KB, MB, GB or rows.")
```

Comment:
Is the error function from a package? Could we explicitly state it 
![image](https://github.com/user-attachments/assets/2a957413-2941-448e-9ef5-ed2145f49065)

Or did you mean `stop()`? Actually, I think you did probably mean `stop()` as that's how we do this elsewhere?